src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs

```diff
@@ -2,35 +2,96 @@
 using System;
 using Microsoft.ServiceBus.Messaging;
 using Orleans.Providers.Streams.Common;
-using Orleans.ServiceBus.Providers;
 using Orleans.Streams;
 
-namespace OrleansServiceBus.Providers.Streams.EventHub
+namespace Orleans.ServiceBus.Providers
 {
-    public abstract class EventHubQueueCache<TCachedMessage> : PooledQueueCache<EventData, TCachedMessage>, IEventHubQueueCache
+    public abstract class EventHubQueueCache<TCachedMessage> : IEventHubQueueCache
         where TCachedMessage : struct
     {
+        protected readonly int defaultMaxAddCount;
+        protected readonly PooledQueueCache<EventData, TCachedMessage> cache;
+        private readonly CachePressureMonitor cachePressureMonitor;
+
         protected IStreamQueueCheckpointer<string> Checkpointer { private set; get; }
 
-        protected EventHubQueueCache(IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
-            : base(cacheDataAdapter, comparer)
+        protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
         {
-            cacheDataAdapter.PurgeAction = Purge;
+            this.defaultMaxAddCount = defaultMaxAddCount;
             Checkpointer = checkpointer;
-            OnPurged = CheckpointOnPurged;
+
+            cache = new PooledQueueCache<EventData, TCachedMessage>(cacheDataAdapter, comparer);
+            cacheDataAdapter.PurgeAction = cache.Purge;
+            cache.OnPurged = CheckpointOnPurged;
+
+            cachePressureMonitor = new CachePressureMonitor();
         }
 
+        protected abstract string GetOffset(TCachedMessage lastItemPurged);
+
+        // cachePressureContribution should be a double between 0-1, indicating how much danger the item is of being removed from the cache.  0 indicating  no danger, 1 indecating removal is imminent.
+        protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
+
         public void Dispose()
         {
-            OnPurged = null;
+            cache.OnPurged = null;
         }
 
-        protected abstract string GetOffset(TCachedMessage lastItemPurged);
-
         private void CheckpointOnPurged(TCachedMessage lastItemPurged)
         {
             Checkpointer.Update(GetOffset(lastItemPurged), DateTime.UtcNow);
         }
+
+        public int MaxAddCount { get { return cachePressureMonitor.IsUnderPressure() ? 0 : defaultMaxAddCount; } }
```

Comment:
the MaxAddCount is part of the stream adapter's public api.  I was trying not to introduce breaking changes, but you're correct, this probably shouldn't be a property.